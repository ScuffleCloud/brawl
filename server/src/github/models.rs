use chrono::{DateTime, Utc};
use octocrab::models::pulls::{MergeableState, ReviewState};
use octocrab::models::{InstallationId, IssueState, RepositoryId, RunId, UserId, WorkflowId};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct User {
    pub login: String,
    pub id: UserId,
}

impl Default for User {
    fn default() -> Self {
        Self {
            login: "user".to_string(),
            id: UserId(0),
        }
    }
}

impl From<octocrab::models::UserProfile> for User {
    fn from(value: octocrab::models::UserProfile) -> Self {
        Self {
            login: value.login,
            id: value.id,
        }
    }
}

impl From<octocrab::models::Author> for User {
    fn from(value: octocrab::models::Author) -> Self {
        Self {
            login: value.login,
            id: value.id,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct Label {
    pub name: String,
}

impl From<octocrab::models::Label> for Label {
    fn from(value: octocrab::models::Label) -> Self {
        Self { name: value.name }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct PrBranch {
    pub label: Option<String>,
    pub ref_field: String,
    pub sha: String,
    pub user: Option<User>,
    pub repo: Option<Repository>,
}

impl From<octocrab::models::pulls::Head> for PrBranch {
    fn from(value: octocrab::models::pulls::Head) -> Self {
        Self {
            label: value.label,
            ref_field: value.ref_field,
            sha: value.sha,
            user: value.user.map(|u| u.into()),
            repo: value.repo.map(|r| r.into()),
        }
    }
}

impl From<octocrab::models::pulls::Base> for PrBranch {
    fn from(value: octocrab::models::pulls::Base) -> Self {
        Self {
            label: value.label,
            ref_field: value.ref_field,
            sha: value.sha,
            user: value.user.map(|u| u.into()),
            repo: value.repo.map(|r| r.into()),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Default)]
pub struct PullRequest {
    pub number: u64,
    pub title: String,
    pub body: String,
    pub state: Option<IssueState>,
    pub mergeable_state: Option<MergeableState>,
    pub merge_commit_sha: Option<String>,
    pub assignees: Vec<User>,
    pub requested_reviewers: Vec<User>,
    pub user: Option<User>,
    pub labels: Vec<Label>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
    pub draft: Option<bool>,
    pub closed_at: Option<DateTime<Utc>>,
    pub merged_at: Option<DateTime<Utc>>,
    pub head: PrBranch,
    pub base: PrBranch,
}

impl From<octocrab::models::pulls::PullRequest> for PullRequest {
    fn from(value: octocrab::models::pulls::PullRequest) -> Self {
        Self {
            number: value.number,
            title: value.title.unwrap_or_default(),
            body: value.body.unwrap_or_default(),
            state: value.state,
            mergeable_state: value.mergeable_state,
            merge_commit_sha: value.merge_commit_sha,
            assignees: value.assignees.into_iter().flatten().map(|a| a.into()).collect(),
            requested_reviewers: value.requested_reviewers.into_iter().flatten().map(|a| a.into()).collect(),
            user: value.user.map(|u| (*u).into()),
            labels: value.labels.into_iter().flatten().map(|l| l.into()).collect(),
            created_at: value.created_at,
            updated_at: value.updated_at,
            draft: value.draft,
            closed_at: value.closed_at,
            merged_at: value.merged_at,
            head: (*value.head).into(),
            base: (*value.base).into(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Default)]
pub struct Review {
    pub user: Option<User>,
    pub state: Option<ReviewState>,
}

impl From<octocrab::models::pulls::Review> for Review {
    fn from(value: octocrab::models::pulls::Review) -> Self {
        Self {
            user: value.user.map(|u| u.into()),
            state: value.state,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct Commit {
    pub sha: String,
    pub tree: Option<String>,
}

impl From<octocrab::models::commits::Commit> for Commit {
    fn from(value: octocrab::models::commits::Commit) -> Self {
        Self {
            sha: value.sha,
            tree: Some(value.commit.tree.sha),
        }
    }
}

impl From<octocrab::models::commits::GitCommitObject> for Commit {
    fn from(value: octocrab::models::commits::GitCommitObject) -> Self {
        Self {
            sha: value.sha,
            tree: Some(value.tree.sha),
        }
    }
}

#[derive(Debug, Clone, serde::Deserialize, PartialEq, Eq, Default)]
#[serde(rename_all = "snake_case")]
pub enum CheckRunStatus {
    #[default]
    Queued,
    InProgress,
    Completed,
    Waiting,
    Requested,
    Pending,
}

#[derive(Debug, Clone, serde::Deserialize, PartialEq, Eq, Default)]
#[serde(rename_all = "snake_case")]
pub enum CheckRunConclusion {
    Success,
    Failure,
    #[default]
    Neutral,
    Cancelled,
    Skipped,
    TimedOut,
    ActionRequired,
}

#[derive(Debug, Clone, serde::Deserialize, PartialEq, Eq, Default)]
pub struct CheckRunEvent {
    pub id: i64,
    pub name: String,
    pub head_sha: String,
    pub html_url: Option<String>,
    pub details_url: Option<String>,
    pub url: String,
    pub started_at: Option<DateTime<Utc>>,
    pub completed_at: Option<DateTime<Utc>>,
    pub status: CheckRunStatus,
    pub conclusion: Option<CheckRunConclusion>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Repository {
    pub id: RepositoryId,
    pub name: String,
    pub owner: User,
    pub default_branch: Option<String>,
}

impl From<octocrab::models::Repository> for Repository {
    fn from(value: octocrab::models::Repository) -> Self {
        Self {
            id: value.id,
            name: value.name,
            owner: value.owner.expect("repository has no owner").into(),
            default_branch: value.default_branch,
        }
    }
}

impl Default for Repository {
    fn default() -> Self {
        Self {
            id: RepositoryId(0),
            name: "repo".to_string(),
            owner: User::default(),
            default_branch: Some("main".to_string()),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Installation {
    pub id: InstallationId,
    pub account: User,
}

impl Default for Installation {
    fn default() -> Self {
        Self {
            id: InstallationId(0),
            account: User::default(),
        }
    }
}

impl From<octocrab::models::Installation> for Installation {
    fn from(value: octocrab::models::Installation) -> Self {
        Self {
            id: value.id,
            account: value.account.into(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct WorkflowRun {
    pub id: RunId,
    pub workflow_id: WorkflowId,
    pub name: String,
    pub head_sha: String,
    pub head_branch: String,
    pub conclusion: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub status: String,
}

impl From<octocrab::models::workflows::Run> for WorkflowRun {
    fn from(value: octocrab::models::workflows::Run) -> Self {
        Self {
            id: value.id,
            workflow_id: value.workflow_id,
            name: value.name,
            head_sha: value.head_sha,
            head_branch: value.head_branch,
            conclusion: value.conclusion,
            created_at: value.created_at,
            updated_at: value.updated_at,
            status: value.status,
        }
    }
}

impl Default for WorkflowRun {
    fn default() -> Self {
        Self {
            id: RunId(0),
            workflow_id: WorkflowId(0),
            name: "workflow".to_string(),
            head_sha: "sha".to_string(),
            head_branch: "branch".to_string(),
            conclusion: None,
            created_at: Utc::now(),
            updated_at: Utc::now(),
            status: "status".to_string(),
        }
    }
}
