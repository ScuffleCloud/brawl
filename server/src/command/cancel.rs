use anyhow::Context;
use diesel::OptionalExtension;
use diesel_async::{AsyncPgConnection, RunQueryDsl};

use super::BrawlCommandContext;
use crate::database::ci_run::CiRun;
use crate::database::pr::Pr;
use crate::github::installation::GitHubRepoClient;
use crate::github::messages;

pub async fn handle<R: GitHubRepoClient>(
    conn: &mut AsyncPgConnection,
    context: BrawlCommandContext<'_, R>,
) -> anyhow::Result<()> {
    Pr::new(&context.pr, context.user.id, context.repo.id())
        .upsert()
        .get_result(conn)
        .await
        .context("update pr")?;

    if let Some(run) = CiRun::active(context.repo.id(), context.pr.number)
        .get_result(conn)
        .await
        .optional()
        .context("fetch ci run")?
    {
        let has_perms = if run.is_dry_run {
            context.repo.can_try(context.user.id).await?
        } else {
            context.repo.can_merge(context.user.id).await?
        };

        if !has_perms {
            tracing::debug!("user does not have permission to do this");
            return Ok(());
        }

        run.cancel(conn, context.repo).await.context("cancel ci run")?;

        context
            .repo
            .send_message(context.pr.number, &messages::error_no_body("Cancelled CI run"))
            .await?;
    }

    Ok(())
}
