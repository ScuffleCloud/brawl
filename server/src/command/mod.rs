use std::str::FromStr;
use std::sync::Arc;

use diesel_async::AsyncPgConnection;
use dry_run::DryRunCommand;
use merge::MergeCommand;
use octocrab::models::pulls::PullRequest;
use octocrab::models::{Author, UserId, UserProfile};

use crate::github::installation::GitHubRepoClient;

mod cancel;
mod dry_run;
mod merge;
mod ping;
mod pr;
mod retry;

pub use pr::PullRequestCommand;

pub enum BrawlCommand {
    DryRun(DryRunCommand),
    Merge(MergeCommand),
    Retry,
    Cancel,
    Ping,
    PullRequest(PullRequestCommand),
}

#[derive(Debug)]
pub struct User {
    pub id: UserId,
    pub login: String,
}

impl From<UserProfile> for User {
    fn from(user: UserProfile) -> Self {
        Self {
            id: user.id,
            login: user.login,
        }
    }
}

impl From<Author> for User {
    fn from(author: Author) -> Self {
        Self {
            id: author.id,
            login: author.login,
        }
    }
}

pub struct BrawlCommandContext<'a, R> {
    pub repo: &'a R,
    pub user: User,
    pub pr: Arc<PullRequest>,
}

impl BrawlCommand {
    pub async fn handle<R: GitHubRepoClient>(
        self,
        conn: &mut AsyncPgConnection,
        context: BrawlCommandContext<'_, R>,
    ) -> anyhow::Result<()> {
        match self {
            BrawlCommand::DryRun(command) => dry_run::handle(conn, context, command).await,
            BrawlCommand::Merge(command) => merge::handle(conn, context, command).await,
            BrawlCommand::Retry => retry::handle(conn, context).await,
            BrawlCommand::Cancel => cancel::handle(conn, context).await,
            BrawlCommand::Ping => ping::handle(conn, context).await,
            BrawlCommand::PullRequest(command) => pr::handle(conn, context, command).await,
        }
    }
}

pub enum BrawlCommandError {
    NoCommand,
    InvalidCommand(String),
    InvalidSyntax(String),
    InvalidArgument(String),
}

impl FromStr for BrawlCommand {
    type Err = BrawlCommandError;

    fn from_str(body: &str) -> Result<Self, Self::Err> {
        let lower = body.to_lowercase();
        let mut splits = lower.split_whitespace();

        let Some(command) = splits
            .find(|s| matches!(*s, "?brawl" | "@brawl" | "/brawl"))
            .and_then(|_| splits.next())
        else {
            return Err(BrawlCommandError::NoCommand);
        };

        match command {
            "merge" => {
                let mut reviewers = Vec::new();
                let mut priority = None;

                for split in splits.by_ref() {
                    if let Some(split) = split.strip_prefix("r=") {
                        if split.is_empty() {
                            tracing::debug!("invalid syntax, reviewer's name cannot be empty");
                            return Err(BrawlCommandError::InvalidSyntax("reviewer's name cannot be empty".into()));
                        }

                        let splits = split.split(',');
                        for reviewer in splits {
                            if reviewer.is_empty() {
                                tracing::debug!("invalid syntax, reviewer's name cannot be empty");
                                return Err(BrawlCommandError::InvalidSyntax("reviewer's name cannot be empty".into()));
                            }

                            reviewers.push(reviewer.to_string());
                        }
                    } else if let Some(split) = split.strip_prefix("p=") {
                        if split.is_empty() {
                            tracing::debug!("invalid syntax, priority cannot be empty");
                            return Err(BrawlCommandError::InvalidSyntax("priority cannot be empty".into()));
                        }

                        let Ok(p) = split.parse() else {
                            tracing::debug!("invalid syntax, priority must be a positive integer");
                            return Err(BrawlCommandError::InvalidSyntax("priority must be a positive integer".into()));
                        };

                        priority = Some(p);
                    } else {
                        break;
                    }
                }

                Ok(BrawlCommand::Merge(MergeCommand { reviewers, priority }))
            }
            "cancel" => Ok(BrawlCommand::Cancel),
            "try" => {
                let mut head_sha = None;
                let mut base_sha = None;

                for next_str in splits {
                    if let Some(head) = next_str
                        .strip_prefix("commit=")
                        .or_else(|| next_str.strip_prefix("c="))
                        .or_else(|| next_str.strip_prefix("head="))
                        .or_else(|| next_str.strip_prefix("h="))
                    {
                        head_sha = Some(head.to_string());
                    } else if let Some(base) = next_str.strip_prefix("base=").or_else(|| next_str.strip_prefix("b=")) {
                        base_sha = Some(base.to_string());
                    } else {
                        break;
                    }
                }

                Ok(BrawlCommand::DryRun(DryRunCommand { head_sha, base_sha }))
            }
            "retry" => Ok(BrawlCommand::Retry),
            "ping" => Ok(BrawlCommand::Ping),
            command => {
                tracing::debug!("invalid command: {}", command);
                Err(BrawlCommandError::InvalidCommand(command.into()))
            }
        }
    }
}
