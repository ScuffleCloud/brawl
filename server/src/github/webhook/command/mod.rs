use std::str::FromStr;
use std::sync::Arc;

use anyhow::Context;
use diesel_async::pooled_connection::bb8;
use diesel_async::{AsyncPgConnection, TransactionManager};
use dry_run::DryRunCommand;
use octocrab::models::pulls::PullRequest;
use octocrab::models::{Author, RepositoryId, UserId, UserProfile};
use pr::PullRequestCommand;
use review::{ReviewAction, ReviewCommand};

use crate::github::config::GitHubBrawlRepoConfig;
use crate::github::installation::InstallationClient;

pub mod cancel;
pub mod dry_run;
pub mod ping;
pub mod pr;
pub mod retry;
pub mod review;
mod utils;

pub enum BrawlCommand {
	DryRun(DryRunCommand),
	Review(ReviewCommand),
	Cancel,
	Retry,
	Ping,
	PullRequest(PullRequestCommand),
}

#[derive(Debug)]
pub struct User {
	pub id: UserId,
	pub login: String,
}

impl From<UserProfile> for User {
	fn from(user: UserProfile) -> Self {
		Self {
			id: user.id,
			login: user.login,
		}
	}
}

impl From<Author> for User {
	fn from(author: Author) -> Self {
		Self {
			id: author.id,
			login: author.login,
		}
	}
}

#[derive(Debug)]
pub struct BrawlCommandContext {
	pub repo_id: RepositoryId,
	pub user: User,
	pub issue_number: u64,
	pub pr: Option<PullRequest>,
	pub config: GitHubBrawlRepoConfig,
}

impl BrawlCommand {
	pub async fn handle(
		self,
		client: &Arc<InstallationClient>,
		database: &bb8::Pool<AsyncPgConnection>,
		context: BrawlCommandContext,
	) -> anyhow::Result<()> {
		let mut conn = database.get().await.context("database get")?;

		diesel_async::AnsiTransactionManager::begin_transaction(&mut *conn)
			.await
			.context("begin transaction")?;

		let result = match self {
			BrawlCommand::DryRun(command) => dry_run::handle(client, &mut *conn, context, command).await,
			BrawlCommand::Review(command) => review::handle(client, &mut *conn, context, command).await,
			BrawlCommand::Cancel => cancel::handle(client, &mut *conn, context).await,
			BrawlCommand::Retry => retry::handle(client, &mut *conn, context).await,
			BrawlCommand::Ping => ping::handle(client, &mut *conn, context).await,
			BrawlCommand::PullRequest(command) => pr::handle(client, &mut *conn, context, command).await,
		};

		if let Err(err) = result {
			if let Err(tx_err) = diesel_async::AnsiTransactionManager::rollback_transaction(&mut *conn)
				.await
				.context("rollback transaction")
			{
				tracing::error!("failed to rollback transaction: {:#}", tx_err);
			}

			Err(err)
		} else {
			diesel_async::AnsiTransactionManager::commit_transaction(&mut *conn)
				.await
				.context("commit transaction")?;
			Ok(())
		}
	}
}

impl FromStr for BrawlCommand {
	type Err = ();

	fn from_str(body: &str) -> Result<Self, Self::Err> {
		let lower = body.to_lowercase();
		let mut splits = lower.split_whitespace();

		while let Some(command) = splits.next() {
			if !command.eq_ignore_ascii_case("?brawl") {
				continue;
			}

			match splits.next().unwrap_or_default() {
				"r+" | "r-" => {
					let mut reviewers = Vec::new();
					let mut priority = None;

					for split in splits.by_ref() {
						if let Some(split) = split.strip_prefix("r=") {
							if split.is_empty() {
								tracing::debug!("invalid syntax, reviewer's name cannot be empty");
								return Err(());
							}

							let splits = split.split(',');
							for reviewer in splits {
								if reviewer.is_empty() {
									tracing::debug!("invalid syntax, reviewer's name cannot be empty");
									return Err(());
								}

								reviewers.push(reviewer.to_string());
							}
						} else if let Some(split) = split.strip_prefix("p=") {
							if split.is_empty() {
								tracing::debug!("invalid syntax, priority cannot be empty");
								return Err(());
							}

							let Ok(p) = split.parse() else {
								tracing::debug!("invalid syntax, priority must be a positive integer");
								return Err(());
							};

							priority = Some(p);
						} else {
							break;
						}
					}

					let action = match command {
						"r+" => ReviewAction::Approve,
						"r-" => ReviewAction::Unapprove,
						_ => unreachable!(),
					};

					return Ok(BrawlCommand::Review(ReviewCommand {
						action,
						reviewers,
						priority,
					}));
				}
				"try" => {
					let commit_sha = splits
						.next()
						.and_then(|s| s.strip_prefix("commit="))
						.map(|commit| commit.to_string());

					return Ok(BrawlCommand::DryRun(DryRunCommand {
						head_sha: commit_sha,
						base_sha: None,
					}));
				}
				"cancel" => {
					return Ok(BrawlCommand::Cancel);
				}
				"retry" => {
					return Ok(BrawlCommand::Retry);
				}
				"ping" => {
					return Ok(BrawlCommand::Ping);
				}
				command => {
					tracing::debug!("invalid command: {}", command);
					return Err(());
				}
			}
		}

		tracing::debug!("no command found");

		Err(())
	}
}
