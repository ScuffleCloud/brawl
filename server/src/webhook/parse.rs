use std::str::FromStr;

use axum::body::Body;
use axum::http::{HeaderMap, HeaderValue, Request, StatusCode};
use hmac::{Hmac, Mac};
use octocrab::models::webhook_events::payload::{
    InstallationWebhookEventAction, IssueCommentWebhookEventAction, RepositoryWebhookEventAction,
};
use octocrab::models::webhook_events::{EventInstallation, WebhookEventPayload, WebhookEventType};
use octocrab::models::{InstallationId, RepositoryId};
use sha2::Sha256;

use crate::command::BrawlCommand;
use crate::github::models::{Installation, Repository, User};

fn verify_gh_signature(headers: &HeaderMap<HeaderValue>, body: &[u8], secret: &str) -> bool {
    let Some(signature) = headers.get("x-hub-signature-256").map(|v| v.as_bytes()) else {
        return false;
    };
    let Some(signature) = signature.get(b"sha256=".len()..).and_then(|v| hex::decode(v).ok()) else {
        return false;
    };

    let mut mac = Hmac::<Sha256>::new_from_slice(secret.as_bytes()).expect("Cannot create HMAC key");
    mac.update(body);
    mac.verify_slice(&signature).is_ok()
}

#[derive(Debug, Clone)]
struct WebhookEvent {
    pub sender: Option<User>,
    pub repository: Option<Repository>,
    pub installtion_id: Option<InstallationId>,
    pub installation: Option<Installation>,
    pub kind: WebhookEventType,
    pub specific: WebhookEventPayload,
}

#[derive(Debug, Clone)]
pub enum WebhookEventAction {
    Command {
        installation_id: InstallationId,
        command: BrawlCommand,
        repo_id: RepositoryId,
        pr_number: u64,
        user: User,
    },
    PullRequest {
        installation_id: InstallationId,
        repo_id: RepositoryId,
        pr_number: u64,
        user: User,
    },
    CheckRun {
        installation_id: InstallationId,
        repo_id: RepositoryId,
        check_run: serde_json::Value,
    },
    DeleteInstallation {
        installation_id: InstallationId,
    },
    AddRepository {
        installation_id: InstallationId,
        repo_id: RepositoryId,
    },
    RemoveRepository {
        installation_id: InstallationId,
        repo_id: RepositoryId,
    },
    UpdateInstallation {
        installation: Installation,
    },
}

fn parse_event(header: String, body: &[u8]) -> Result<Vec<WebhookEventAction>, (StatusCode, String)> {
    let kind: WebhookEventType =
        serde_json::from_value(serde_json::Value::String(header)).map_err(|e| (StatusCode::BAD_REQUEST, e.to_string()))?;

    // Intermediate structure allows to separate the common fields from
    // the event specific one.
    #[derive(serde::Deserialize)]
    struct Intermediate {
        sender: Option<octocrab::models::Author>,
        repository: Option<octocrab::models::Repository>,
        #[allow(unused)]
        organization: Option<octocrab::models::orgs::Organization>,
        installation: Option<octocrab::models::webhook_events::EventInstallation>,
        #[serde(flatten)]
        specific: serde_json::Value,
    }

    let Intermediate {
        sender,
        repository,
        organization: _,
        installation,
        mut specific,
    } = serde_json::from_slice::<Intermediate>(body).map_err(|e| (StatusCode::BAD_REQUEST, e.to_string()))?;

    // Bug: OctoCrab wrongly requires the pusher to have an email
    // Remove when https://github.com/XAMPPRocky/octocrab/issues/486 is fixed
    if kind == WebhookEventType::Push {
        if let Some(pusher) = specific.get_mut("pusher") {
            if let Some(email) = pusher.get_mut("email") {
                if email.is_null() {
                    *email = serde_json::Value::String("".to_owned())
                }
            }
        }
    }

    let specific = kind
        .parse_specific_payload(specific)
        .map_err(|e| (StatusCode::BAD_REQUEST, e.to_string()))?;

    let actions = parse_webhook(WebhookEvent {
        sender: sender.map(|s| s.into()),
        repository: repository.map(|r| r.into()),
        installtion_id: match &installation {
            Some(EventInstallation::Full(installation)) => Some(installation.id),
            Some(EventInstallation::Minimal(installation)) => Some(installation.id),
            None => None,
        },
        installation: match installation {
            Some(EventInstallation::Full(installation)) => Some((*installation).into()),
            _ => None,
        },
        kind,
        specific,
    });

    Ok(actions)
}

fn parse_webhook(mut event: WebhookEvent) -> Vec<WebhookEventAction> {
    let mut actions = vec![];

    if let Some(installation) = event.installation {
        actions.push(WebhookEventAction::UpdateInstallation { installation });
    }

    let Some(installation_id) = event.installtion_id else {
        tracing::warn!("event does not have an installation id: {:?}", event.kind);
        return actions;
    };

    match event.specific {
        WebhookEventPayload::Installation(install_event) => match install_event.action {
            InstallationWebhookEventAction::Deleted | InstallationWebhookEventAction::Suspend => {
                actions.push(WebhookEventAction::DeleteInstallation { installation_id });
            }
            _ => {}
        },
        WebhookEventPayload::InstallationRepositories(event) => {
            for repo in event.repositories_added {
                actions.push(WebhookEventAction::AddRepository {
                    installation_id,
                    repo_id: repo.id,
                });
            }

            for repo in event.repositories_removed {
                actions.push(WebhookEventAction::RemoveRepository {
                    installation_id,
                    repo_id: repo.id,
                });
            }
        }
        WebhookEventPayload::Repository(repo_event) => {
            let Some(repo) = event.repository else {
                return actions;
            };

            match repo_event.action {
                RepositoryWebhookEventAction::Deleted | RepositoryWebhookEventAction::Archived => {
                    actions.push(WebhookEventAction::RemoveRepository {
                        installation_id,
                        repo_id: repo.id,
                    });
                }
                _ => {
                    actions.push(WebhookEventAction::AddRepository {
                        installation_id,
                        repo_id: repo.id,
                    });
                }
            }
        }
        WebhookEventPayload::PullRequest(mut pull_request_event) => {
            let Some(repo_id) = event.repository.as_ref().map(|r| r.id).or(pull_request_event
                .pull_request
                .repo
                .as_deref()
                .map(|r| r.id))
            else {
                return actions;
            };

            let Some(user) = event
                .sender
                .take()
                .or_else(|| pull_request_event.pull_request.user.take().map(|u| (*u).into()))
            else {
                return actions;
            };

            actions.push(WebhookEventAction::PullRequest {
                installation_id,
                repo_id,
                pr_number: pull_request_event.pull_request.number,
                user,
            });
        }
        WebhookEventPayload::IssueComment(issue_comment_event)
            if issue_comment_event.action == IssueCommentWebhookEventAction::Created
                && issue_comment_event.issue.pull_request.is_some() =>
        {
            let Some(body) = issue_comment_event.comment.body.as_ref() else {
                return actions;
            };

            let Ok(command) = BrawlCommand::from_str(body) else {
                return actions;
            };

            let Some(repo) = event.repository else {
                return actions;
            };

            actions.push(WebhookEventAction::Command {
                installation_id,
                command,
                repo_id: repo.id,
                pr_number: issue_comment_event.issue.number,
                user: issue_comment_event.comment.user.into(),
            });
        }
        WebhookEventPayload::CheckRun(check_run_event) => {
            let Some(repo) = event.repository else {
                return actions;
            };

            actions.push(WebhookEventAction::CheckRun {
                installation_id,
                repo_id: repo.id,
                check_run: check_run_event.check_run,
            });
        }
        _ => {}
    }

    actions
}

pub async fn parse_from_request(
    request: Request<Body>,
    secret: &str,
) -> Result<Vec<WebhookEventAction>, (StatusCode, String)> {
    let (parts, body) = request.into_parts();
    let Some(header) = parts.headers.get("X-GitHub-Event").and_then(|v| v.to_str().ok()) else {
        return Err((StatusCode::BAD_REQUEST, "Missing X-GitHub-Event header".to_string()));
    };

    let Ok(body) = axum::body::to_bytes(body, 1024 * 1024 * 10).await else {
        return Err((StatusCode::BAD_REQUEST, "Failed to read body".to_string()));
    };

    if !verify_gh_signature(&parts.headers, &body, secret) {
        return Err((StatusCode::UNAUTHORIZED, "Invalid signature".to_string()));
    }

    parse_event(header.to_owned(), &body)
}

#[cfg(test)]
#[cfg_attr(coverage_nightly, coverage(off))]
mod tests {
    use super::*;

    #[test]
    fn test_parse_event() {
        let event = parse_event("check_run".to_string(), include_bytes!("mock/webhook.check_run.created.json"))
            .expect("event parsed");
        insta::assert_debug_snapshot!("webhook.check_run.created", event);

        let event = parse_event(
            "check_run".to_string(),
            include_bytes!("mock/webhook.check_run.completed.json"),
        )
        .expect("event parsed");
        insta::assert_debug_snapshot!("webhook.check_run.completed", event);

        let event = parse_event("push".to_string(), include_bytes!("mock/webhook.push.json")).expect("event parsed");
        insta::assert_debug_snapshot!("webhook.push", event);

        let event = parse_event(
            "pull_request".to_string(),
            include_bytes!("mock/webhook.pull_request.synchronize.json"),
        )
        .expect("event parsed");
        insta::assert_debug_snapshot!("webhook.pull_request.synchronize", event);

        let event = parse_event(
            "issue_comment".to_string(),
            include_bytes!("mock/webhook.issue_comment.created.json"),
        )
        .expect("event parsed");
        insta::assert_debug_snapshot!("webhook.issue_comment.created", event);

        let event = parse_event("delete".to_string(), include_bytes!("mock/webhook.delete.json")).expect("event parsed");
        insta::assert_debug_snapshot!("webhook.delete", event);

        let event = parse_event("create".to_string(), include_bytes!("mock/webhook.create.json")).expect("event parsed");
        insta::assert_debug_snapshot!("webhook.create", event);

        let event = parse_event(
            "pull_request_review".to_string(),
            include_bytes!("mock/webhook.pull_request_review.submitted.json"),
        )
        .expect("event parsed");
        insta::assert_debug_snapshot!("webhook.pull_request_review.submitted", event);

        let event = parse_event(
            "issue_comment".to_string(),
            include_bytes!("mock/webhook.issue_comment.created-brawl-cmd.json"),
        )
        .expect("event parsed");
        insta::assert_debug_snapshot!("webhook.issue_comment.created-brawl-cmd", event);

        let event = parse_event(
            "installation_repositories".to_string(),
            include_bytes!("mock/webhook.installation_repositories.added.json"),
        )
        .expect("event parsed");
        insta::assert_debug_snapshot!("webhook.installation_repositories.added", event);

        let event = parse_event(
            "installation_repositories".to_string(),
            include_bytes!("mock/webhook.installation_repositories.removed.json"),
        )
        .expect("event parsed");
        insta::assert_debug_snapshot!("webhook.installation_repositories.removed", event);

        let event = parse_event(
            "repository".to_string(),
            include_bytes!("mock/webhook.repository.deleted.json"),
        )
        .expect("event parsed");
        insta::assert_debug_snapshot!("webhook.repository.deleted", event);

        let event = parse_event(
            "repository".to_string(),
            include_bytes!("mock/webhook.repository.created.json"),
        )
        .expect("event parsed");
        insta::assert_debug_snapshot!("webhook.repository.created", event);

        let event = parse_event(
            "installation".to_string(),
            include_bytes!("mock/webhook.installation.suspend.json"),
        )
        .expect("event parsed");
        insta::assert_debug_snapshot!("webhook.installation.suspend", event);

        let event = parse_event(
            "installation".to_string(),
            include_bytes!("mock/webhook.installation.unsuspend.json"),
        )
        .expect("event parsed");
        insta::assert_debug_snapshot!("webhook.installation.unsuspend", event);
    }

    #[test]
    fn test_verify_gh_signature() {
        let mut mac = Hmac::<Sha256>::new_from_slice("secret".as_bytes()).expect("Cannot create HMAC key");
        mac.update(b"body");
        let signature = mac.finalize().into_bytes().to_vec();
        let signature = hex::encode(signature);
        let signature = format!("sha256={}", signature);

        let mut headers = HeaderMap::new();
        headers.insert("X-Hub-Signature-256", signature.parse().unwrap());
        assert!(verify_gh_signature(&headers, b"body", "secret"));
    }

    #[tokio::test]
    async fn test_parse_from_request() {
        let data = include_bytes!("mock/webhook.push.json");

        let mut mac = Hmac::<Sha256>::new_from_slice("secret".as_bytes()).expect("Cannot create HMAC key");
        mac.update(data);
        let signature = mac.finalize().into_bytes().to_vec();
        let signature = hex::encode(signature);
        let signature = format!("sha256={}", signature);

        let request = Request::builder()
            .header("X-Hub-Signature-256", signature)
            .header("X-GitHub-Event", "push")
            .body(Body::from(data.to_vec()))
            .unwrap();

        parse_from_request(request, "secret").await.expect("event parsed");
    }

    #[tokio::test]
    async fn test_parse_from_request_invalid_signature() {
        let data = include_bytes!("mock/webhook.push.json");
        let request = Request::builder()
            .header("X-GitHub-Event", "push")
            .body(Body::from(data.to_vec()))
            .unwrap();
        let err = parse_from_request(request, "secret").await.expect_err("error");
        assert_eq!(err, (StatusCode::UNAUTHORIZED, "Invalid signature".to_string()));
    }

    #[tokio::test]
    async fn test_parse_from_request_missing_header() {
        let data = include_bytes!("mock/webhook.push.json");
        let request = Request::builder().body(Body::from(data.to_vec())).unwrap();
        let err = parse_from_request(request, "secret").await.expect_err("error");
        assert_eq!(err, (StatusCode::BAD_REQUEST, "Missing X-GitHub-Event header".to_string()));
    }

    #[tokio::test]
    async fn test_parse_from_request_invalid_body() {
        let body = Body::from(vec![0; 1024 * 1024 * 20]);
        let request = Request::builder().header("X-GitHub-Event", "push").body(body).unwrap();
        let err = parse_from_request(request, "secret").await.expect_err("error");
        assert_eq!(err, (StatusCode::BAD_REQUEST, "Failed to read body".to_string()));
    }
}
