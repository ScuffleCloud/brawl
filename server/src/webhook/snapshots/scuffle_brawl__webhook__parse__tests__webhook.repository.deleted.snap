---
source: server/src/webhook/parse.rs
expression: event
snapshot_kind: text
---
[
    RemoveRepository {
        installation_id: InstallationId(
            58286822,
        ),
        repo_id: RepositoryId(
            907533428,
        ),
    },
]
